addons/udes_advanced_packaging/models/stock_picking.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+from collections import defaultdict
+
+from odoo import models, _
+from odoo.exceptions import ValidationError
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _should_reserve_full_packages(self):
+        """Method to determine if picking should reserve entire packages"""
+        self.picking_type_id.ensure_one()
+        return self.picking_type_id.u_reserve_as_packages
+
+    def _reserve_full_packages(self):
+        """
+        If the picking type of the picking in self has full package
+        reservation enabled, whole packages will be reserved.
+        """
+        MoveLine = self.env["stock.move.line"]
+
+        # do not reserve full packages when bypass_reserve_full packages
+        # is set in the context as True
+        if not self.env.context.get("bypass_reserve_full_packages"):
+            pickings = self.filtered(lambda p: p._should_reserve_full_packages())
+            pickings.move_line_ids.package_id
+
+            for picking in pickings:
+                remaining_qtys = defaultdict(int)
+
+                # get all packages
+                packages = picking.move_line_ids.package_id
+                for package in packages:
+                    move_lines = picking.move_line_ids.filtered(
+                        lambda ml: ml.package_id == package
+                    )
+                    pack_products = frozenset(package._get_all_products_quantities().items())
+                    mls_products = frozenset(move_lines._get_all_products_quantities().items())
+                    if pack_products != mls_products:
+                        # move_lines do not match the quants
+                        pack_mls = MoveLine.search(
+                            [
+                                ("package_id", "child_of", package.id),
+                                ("state", "not in", ["done", "cancel"]),
+                            ]
+                        )
+                        other_pickings = pack_mls.picking_id - picking
+                        if other_pickings:
+                            raise ValidationError(
+                                _("The package is reserved in other pickings: %s")
+                                % ",".join(other_pickings.mapped("name"))
+                            )
+
+                        quants = package._get_contained_quants()
+                        for product, qty in quants.group_quantity_by_product(
+                            only_available=True
+                        ).items():
+                            remaining_qtys[product] += qty
+                if remaining_qtys:
+                    # Context variables:
+                    # - filter the quants used in _create_moves() to be
+                    # the ones of the packages to be completed
+                    # - add bypass_reserve_full_packages at the context
```

Comment:
Move this comment to be above `picking.with_context(bypass_reserve_full_packages=True).action_assign()`. Also change `_create_moves` to `_create_move`.